core/templates/dev/head/domain/topic/TopicRightsObjectFactory.ts

```diff
@@ -17,87 +17,92 @@
-      return this._canPublishTopic;
-    };
-
-    // Sets _isPublished to true only if the user can publish the
-    // corresponding topic.
-    TopicRights.prototype.markTopicAsPublished = function() {
-      if (this._canPublishTopic) {
-        this._published = true;
-      } else {
-        throw new Error('User is not allowed to publish this topic.');
-      }
-    };
+  constructor(
+      published: boolean, canPublishTopic: boolean, canEditTopic: boolean) {
+    this._published = published;
+    this._canPublishTopic = canPublishTopic;
+    this._canEditTopic = canEditTopic;
+  }
 
-    // Sets _isPublished to false if user can unpublish the topic.
-    TopicRights.prototype.markTopicAsUnpublished = function() {
-      if (this._canPublishTopic) {
-        this._published = false;
-      } else {
-        throw new Error('User is not allowed to unpublish this topic.');
-      }
-    };
+  canEditTopic() {
+    return this._canEditTopic;
+  }
+  isPublished() {
+    return this._published;
+  }
+  canPublishTopic() {
+    return this._canPublishTopic;
+  }
+  canEditName() {
+    return this._canPublishTopic;
+  }
+  markTopicAsPublished() {
+    if (this._canPublishTopic) {
+      this._published = true;
+    } else {
+      throw new Error('User is not allowed to publish this topic.');
+    }
+  }
+  markTopicAsUnpublished() {
+    if (this._canPublishTopic) {
+      this._published = false;
+    } else {
+      throw new Error('User is not allowed to unpublish this topic.');
+    }
+  }
+  // Reassigns all values within this topic to match the existing
+  // topic rights. This is performed as a deep copy such that none of the
+  // internal, bindable objects are changed within this topic rights.
+  copyFromTopicRights(
+      otherTopicRights: {
+        isPublished: () => boolean;
+        canEditTopic: () => boolean;
+        canPublishTopic: () => boolean;
+      }) {
+    this._published = otherTopicRights.isPublished();
+    this._canEditTopic = otherTopicRights.canEditTopic();
+    this._canPublishTopic = otherTopicRights.canPublishTopic();
+  }
+}
 
-    // This function takes a JSON object which represents a backend
-    // topic python dict.
-    // TODO (ankita240796) Remove the bracket notation once Angular2 gets in.
-    /* eslint-disable dot-notation */
-    TopicRights['createFromBackendDict'] = function(topicRightsBackendObject) {
-    /* eslint-enable dot-notation */
-      return new TopicRights(
-        topicRightsBackendObject.published,
-        topicRightsBackendObject.can_publish_topic,
-        topicRightsBackendObject.can_edit_topic
-      );
-    };
+@Injectable({
+  providedIn: 'root'
+})
+export class TopicRightsObjectFactory {
+  // This function takes a JSON object which represents a backend
+  // topic python dict.
+  createFromBackendDict(
+      topicRightsBackendObject: {
+        published: boolean;
+        canPublishTopicBool: boolean;
+        canEditTopicBool: boolean;
```

Comment:
I had to change `can_publish_topic` and similarly `can_edit_topic` since they were failing the camelCase rule once they were converted to classes. Further `canPublishTopic` can not be the name since that name is already taken up by a member function.